### PR TITLE
[5.4] Fix: Showon fails open when source field does not exist (hidden by default instead)

### PR DESCRIPTION
**Pull request by coolcat-creations**
This pull request fixes an issue in Joomla’s showon implementation where dependent fields are always shown, even when the referenced source field does not exist in the DOM.
This “fail-open” behaviour can lead to incorrect form states, invalid UI logic, and unintended visibility of fields.

**Problem**
Currently, if a field contains something like:
`showon="nonexistent_field:1"`

and nonexistent_field is not present in the form:

1. Joomla does not register any origin fields,
2. no event listeners are attached,
3. and the dependent field is shown by default.

This is counter-intuitive and causes issues especially in:

1. Custom Fields, when a controlling field is deleted or renamed
2. Complex XML forms, especially in components or modules using conditional configurations
3. Subforms, where dynamic names change and conditions may temporarily point to a non-existing field

This behaviour is likely an oversight rather than an intentional feature.

**Expected Behaviour**

If the source field defined in showon does not exist, then the condition cannot be fulfilled, therefore the dependent field should be hidden by default.

**Solution**

This PR adds a simple and safe check in Showon (build/media_source/system/js/showon.es6.js):

After collecting all origin fields for a given showon condition:

if origin.length === 0,
→ the dependent field is immediately hidden,
→ and the joomla:showon-hide event is fired.

This changes the behaviour to Fail closed instead of fail open
(a missing condition hides a field rather than incorrectly showing it)

The fix does not modify the data structure, PHP FormHelper logic, or any API and is fully backward compatible unless someone relied on the previous incorrect behaviour (unlikely).

**How to Test**

Please test both Custom Fields and standard XML JForm definitions.

**1. Test Using Custom Fields**

A) Prepare the fields

Go to:
Content → Fields

Create a field:

Field A:

Type: List
Name: controller
Options

List values | Text* | Value |
___________ | Yes _ | 1 |
___________ | No _ | 0 |

Create another field:

Field B:

Type: Text
Name: dependent
Showon: `controller:1`

B) Test the normal behaviour

Edit an article.

Set “Controller” to "Yes" → Field B should appear.
Set “Controller” to "No" → Field B should hide.

C) Test the missing-source behaviour

Unpublish or Delete Field A (the controller field)

Open the article again.

Expected result after applying the PR:

“Dependent” field is hidden by default

No JS errors occur

Switching values of other fields does not accidentally reveal it

Current behaviour (before PR):

Field B is incorrectly visible.

**2. Test Using XML Form Definitions**

A) Create a simple XML form (administrator or component)

Example snippet: `<field
    name="dependent"
    type="text"
    label="Dependent Field"
    showon="nonexistent:1"
/>
`

B) Load the form in the backend (module, component, plugin etc.)
Expected result after PR:

The “dependent” field is hidden because the source field does not exist.

Current behaviour:

The field is incorrectly shown.

3. Test With Multiple Conditions

`showon="controller:1[AND]nonexistent:2"
`

Expected after PR:

Even if controller exists, the missing nonexistent should invalidate the entire condition, hiding the field.

**Backward Compatibility**
This change should not break valid configurations.
The only difference occurs when showon references a field that:

does not exist,
was renamed,
was deleted,

or is dynamically removed in subforms.

In these cases the new behaviour is more correct and prevents accidental visibility.

This PR makes the showon behaviour more predictable, safer, and consistent by hiding dependent fields when the source field is missing — a long-standing edge case that often leads to incorrect UI states.

### CODE REVIEW

**Comment by ThomasFinnern** [comment]
**1. Test Using Custom Fields**
...
Options: Yes=1, No=0 
  => The option description can be improved.

Add two lines in "List Values" like blow

List values | Text* | Value |
___________ | Yes _ | 1  |
___________ | No _ | 0  |
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/46488">issues.joomla.org/tracker/joomla-cms/46488</a>.</sub>

**Comment by coolcat-creations** [comment]
@ThomasFinnern changed, thank you - did you test it?

**Comment by ThomasFinnern** [comment]
I tried to get there but failed due to lack of my test knowledge. 
I exchanged the file in build/... and used npm ci. but the result in fileds was the same.
So i will try again
Anyhow some test cases which you may use in an *.xml file using "controller2"
```
<field name="controller2" type="radio" label="controller2" default="1" ><option value="0">JNO</option><option value="1">JYES</option></field>

<field name="dependent01" type="text" label="Dependent Field 01" default="!!! Should be shown always !!!"  />
<field name="dependent02" type="text" label="Dependent Field 02" default="!!! Should be shown !!!" showon="controller2:1" />
<field name="dependent03" type="text" label="Dependent Field 03" default="!!! Should not be shown !!!" showon="nonexistent:1" />
<field name="dependent04" type="text" label="Dependent Field 04" default="!!! Should not be shown !!!" showon="controller2:1[AND]nonexistent:2" />
<field name="dependent05" type="text" label="Dependent Field 05" default="!!! Should not be shown !!!" showon="controller2:1[AND]nonexistent:2" />
<field name="dependent06" type="text" label="Dependent Field 06" default="!!! Should not be shown !!!" showon="nonexistent:2[AND]controller2:1" />
<field name="dependent07" type="text" label="Dependent Field 07" default="!!! Should be shown !!!" showon="nonexistent:2[OR]controller2:1" />
<field name="dependent08" type="text" label="Dependent Field 08" default="!!! Should be shown !!!" showon="controller2:1[OR]nonexistent:2" />
<field name="dependent09" type="text" label="Dependent Field 09" default="!!! Should be shown !!!" showon="controller2:1[OR]controller2:1" />
```

**Comment by bembelimen** [comment]
Yes, your code + the extra comment

**Comment by LadySolveig** [comment]
Somehow, information seems to have been lost along the way to @bembelimen . We agreed in the maintainer team that a warning would definitely be useful.
Thank you for the effort you are putting into improving things here.

The potential problem here arises at the latest when showon is used with a negative condition. Then, unfortunately, the code may have exactly the opposite effect to what you want to achieve and lead to even more confusion.

In the core itself, it should probably remain a simple warning and dispatching the event so that even less experienced users or developers can quickly become aware of the problem and recognise their misconfiguration.

However, if you keep the event in there, you can then add an Eventlistener and hide it in according to your specific needs.

File: build/media_source/system/js/showon.es6.js [hunk 1]
```diff
@@ -86,6 +86,19 @@ class Showon {
               }
             });
           }
+          if (this.fields[showonData[0].field] && !this.fields[showonData[0].field].origin.length) {
+            field.classList.add('hidden');
+            field.dispatchEvent(new CustomEvent('joomla:showon-hide', {
+              bubbles: true,
+            }));
```
Comment (LadySolveig) [inline build/media_source/system/js/showon.es6.js]:
```suggestion
           field.dispatchEvent(new CustomEvent('joomla:showon-hide', {
              detail: {reason: 'missingParent'}
              bubbles: true,
            }));
```

Just a suggestion.
Related comment: https://github.com/joomla/joomla-cms/pull/46488#issuecomment-3617180950
What do you think, @Fedik  @dgrammatiko  would it make sense to have a separate event here?

Comment (Fedik) [reply build/media_source/system/js/showon.es6.js]: I would keep it minimal. Without extra events.
If you really need to have some flag about missing parent then can do kind of:
```javascript
field.dispatchEvent(new CustomEvent('joomla:showon-hide', {
  detail: {missingParent: true}
  bubbles: true,
}));
```
Or:
```javascript
detail: {reason: 'missingParent'}
```

Comment (LadySolveig) [reply build/media_source/system/js/showon.es6.js]: Nice thanks for the suggestion.

You would then leave `field.classList.add(“hidden”);` in place?
Even if that means that in a condition such as, for example, `foo!:` or `foo!:1` are we more likely risk the undesirable side effect of data loss until the user notices it?

Otherwise, I find the event name somewhat confusing.

I don't think it will ever fit in any direction anyway. The main question is in which direction it could potentially cause more problems. 🙃

Comment (Fedik) [reply build/media_source/system/js/showon.es6.js]: Personalty I would not hide the field, only show the warning about misconfiguration.
But if people want we can hide.

There should not be any data lose because it just visual thing and nothing to do with field values.

Comment (LadySolveig) [reply build/media_source/system/js/showon.es6.js]: Just wasn't sure if I had understood you correctly, but that basically confirms what we discussed earlier.

**Comment by LadySolveig** [comment]
Yes, it really isn't easy, I understand what you mean.

I'll give it a try with your example.
If you have the opposite case and, for example, you have the cuisine field for all of them and only have one branch to which this does not apply. 

showon = branches:!retail

then it would never be shown, even though there is a greater potential here that it should be displayed. And by the time you notice, you have potential "data loss" because you will not fill in a field that you cannot see.

File: build/media_source/system/js/showon.es6.js [hunk 2]
```diff
@@ -86,6 +86,19 @@ class Showon {
               }
             });
           }
+          if (this.fields[showonData[0].field] && !this.fields[showonData[0].field].origin.length) {
+           field.dispatchEvent(new CustomEvent('joomla:showon-hide', {
+              detail: {reason: 'missingParent'}
```
Comment (dgrammatiko) [inline build/media_source/system/js/showon.es6.js]:
```suggestion
              detail: {reason: 'missingParent'},
```

**Review by dgrammatiko — DISMISSED** [review]

**Comment by richard67** [comment]
> No, not from my side. Everything seems to be as it should be now. @richard67 @muhme

@LadySolveig As usual, the PR needs 2 successful human tests after it has been changed. Nothing RMs can do here before there are 2 tests.

**Comment by LadySolveig** [comment]
Just FYI, that in my opinion would be ready to test. But as you mentioned, I'm not the RM. :)

**Comment by ceford** [comment]
I really ought to give my test result as a Fail. I use a joomla-cms clone for testing, in this case with 5.4-dev checked out. This morning I updated my clone (git pull, composer install, npm ci) and then applied the patch and ran npm run build:js

With the Controller field unpublished the Dependent field appears in the article edit form. It is not clear from the discussion whether that is intended.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/46488">issues.joomla.org/tracker/joomla-cms/46488</a>.</sub>
